open_spiel/algorithms/cfr.cc

```diff
@@ -123,23 +136,41 @@ ActionsAndProbs CFRCurrentPolicy::GetStatePolicyFromInformationStateValues(
 
 CFRSolverBase::CFRSolverBase(const Game& game, bool alternating_updates,
                              bool linear_averaging, bool regret_matching_plus)
-    : game_(game),
+    : game_(game.Clone()),
       root_state_(game.NewInitialState()),
-      root_reach_probs_(game_.NumPlayers() + 1, 1.0),
+      root_reach_probs_(game_->NumPlayers() + 1, 1.0),
       regret_matching_plus_(regret_matching_plus),
       alternating_updates_(alternating_updates),
       linear_averaging_(linear_averaging),
       chance_player_(game.NumPlayers()) {
-  if (game_.GetType().dynamics != GameType::Dynamics::kSequential) {
+  if (game_->GetType().dynamics != GameType::Dynamics::kSequential) {
     SpielFatalError(
         "CFR requires sequential games. If you're trying to run it "
         "on a simultaneous (or normal-form) game, please first transform it "
         "using turn_based_simultaneous_game.");
   }
-
   InitializeInfostateNodes(*root_state_);
 }
 
+CFRSolverBase::CFRSolverBase(std::shared_ptr<const Game> game,
```

Comment:
Sorry, I'm missing why we need this extra constructor, can you explain?

Reply:
The only difference with the existing one is that it also accepts `iteration` (and that it accepts the game instance as a shared_ptr but this is not really relevant). If we remove this constructor I believe the only other alternative is to create a setter for `iteration_`. Rather than creating setters for fields that get deserialized and need to be set on new instances after deserialization, I created new constructors. Maybe there is a better approach?

Reply:
Please note that `InitializeInfostateNodes()` is not called in this constructor since the table is expected to be deserialized into an empty table later (using the `DeserializeCFRInfoStateValuesTable()` method).